backend/server/routers/courses.py

```diff
@@ -178,3 +178,35 @@ def unselectCourse(userData: UserData, lockedCourses: list, unselectedCourse: st
     affectedCourses = User(fixUserData(userData.dict())).unselect_course(unselectedCourse, lockedCourses)
 
     return {'affected_courses': affectedCourses}
+
+def coursesUnlockedWhenTaken(userData: UserData, courseToBeTaken: str):
+    """ Returns all courses which are unlocked when given course is taken """
+    
+    ## define the user object with user data
+    user = User(fixUserData(userData.dict()))
+    
+    ## initial state
+    courses_initially_unlocked = getAllUnlocked(userData)
+    
+    ## add course to the user
+    courseToAdd = {courseToBeTaken: (getCourse(courseToBeTaken), None)}
+    user.add_courses(courseToAdd)
+    
+    ## final state
+    courses_now_unlocked = getAllUnlocked(user.get_user_data())
+    
+    ## compare 
+    before = []
+    after = []
+    course_keys = list(courses_initially_unlocked.keys())
+    for course in course_keys:
```

Comment:
this and the next should be list comprensions.
or better yet, make them explicitly sets https://www.w3schools.com/python/python_sets.asp. Using this will probably make this code incredibly elegant.